ursabot/builders.py

```diff
@@ -0,0 +1,345 @@
+import copy
+import toolz
+
+from buildbot import interfaces
+from buildbot.plugins import util
+
+from .steps import (ShellCommand, SetPropertiesFromEnv,
+                    Ninja, SetupPy, CMake, PyTest, Mkdir, Pip, GitHub)
+
+
+class BuildFactory(util.BuildFactory):
+
+    def clone(self):
+        return copy.deepcopy(self)
+
+    def add_step(self, step):
+        return super().addStep(step)
+
+    def add_steps(self, steps):
+        return super().addSteps(steps)
+
+    def prepend_step(self, step):
+        self.steps.insert(0, interfaces.IBuildStepFactory(step))
+
+
+class Builder(util.BuilderConfig):
+
+    tags = tuple()
+    steps = tuple()
+    properties = None
+    default_properties = None
+
+    def __init__(self, name=None, steps=None, factory=None, workers=None,
+                 tags=None, properties=None, default_properties=None,
+                 **kwargs):
+        if isinstance(steps, (list, tuple)):
+            # replace the class' steps
+            steps = steps
+        elif steps is None:
+            steps = self.steps
+        else:
+            raise TypeError('Steps must be a list')
+
+        if isinstance(tags, (list, tuple)):
+            # append to the class' tag list
+            tags = list(filter(None, toolz.concat([self.tags, tags])))
+        elif tags is not None:
+            raise TypeError('Tags must be a list')
+
+        factory = factory or BuildFactory(steps)
+        properties = toolz.merge(properties or {}, self.properties or {})
+        default_properties = toolz.merge(default_properties or {},
+                                         self.default_properties or {})
+
+        workernames = None if workers is None else [w.name for w in workers]
+
+        return super().__init__(name=name, tags=tags, properties=properties,
+                                defaultProperties=default_properties,
+                                workernames=workernames, factory=factory,
+                                **kwargs)
+
+
+# prefer GitHub over Git step
```

Comment:
Why?

Reply:
Because it handles the pr/merge reference magic, thus it checks out the merged branch